### ISSUE TITLE
Improve AlsDev's Environment window line-editing

### ISSUE THREAD

**Issue by kenbowen** [AppliedLogicSystems/ALSProlog#74]
Alsdev is a Gui program, so command-line services aren’t available to it.  AlsDev's Environment window line-editing is quite limited, so it ought to be improved.

[Per Chuck]
Tk comes with a fairly advanced console window (editing/history/completion/color-syntax). You can see it in action with Wish:

open /System/Library/Frameworks/Tk.framework/Versions/8.5/Resources/Wish.app

The wish console code could be cherry-picked to improve the AlsDev console. Source code: http://core.tcl.tk/tk/artifact/29cc6e699359401f  (as opposed to trying to directly use the wish console, though that would be a possibility too.)

**Comment by kenbowen** [comment]
I’ve spent time on this issue.  The minimum functionality to be desired would be to be able to move Left (arrow) zero or more characters, and then delete (BackSpace over) zero or more characters, and/or insert zero or more characters, including paste; but it would be desirable that Left and BackSpace don’t consume or pass over the prompt on the line.  

The first part of this is achievable by modifying the line
	bind $WinPath <KeyPress> "listener.check_at_end .topals"
in proc set_top_bindings in alsdev.tcl.  However, protecting the prompt from Left, BackSpace has proved difficult.  The problem is that I seem to be unable to suppress <Keypress> for Left and BackSpace, so that these KeyPresses insert the characters, and the prompt is unprotected, and either erased or crossed over.

I’ve attempted to adapt the code from 
	tk8.6.5/library/console.tcl
in ::tk::ConsoleBind at lines 409ff, but have been unsuccessful so far. I’ve gone so far as to essentially replace the code in proc set_top_bindings in alsdev.tcl by the code in 
::tk::ConsoleBind, without success.  

I’m afraid the essential experiment is to cut out all *.Text components (console, debugger, file edits), and then try to build up a console following how it is done in tk8.6.5/library/console.tcl, afterwards re-introducing debugger, file edits, etc.  I’m going to set this aside for the present.

**Comment by kenbowen** [comment]
When organizing what I’ve done so far to be a reasonable branch to check in (i74_alsdev-line-edit), I tried out something bizarre:  I ran proc alsdev.tcl > set_top_bindings  with an empty body.  Amazingly, the basic console input worked just as it has been (with the exception of special bindings like <Control-d>, <Control-u>, and the binding of <KeyPress> to ”listener.check_at_end .topals”, etc).  Then I commented out the initialization call to set_top_bindings:
	builtins/blt_dvsh.pro > alsdev(Shared, ALS_IDE_Mgr):
	    tcl_call(shl_tcli,
                [set_top_bindings,'.topals.text',shl_tk_in_win,WaitVar,DataVar],_)
And the console still works as before (with the same exceptions).

First observation:  This is a bit of a mess.
Next: The tangle is partially involved with the WaitVariable machinery.  In sio.pro, we see (pbi_write added):
wait_data(tcltk, Stream, Call)
        :-!,
        stream_addl1(Stream, Alias),
        stream_addl2(Stream, ti(Interp,WaitVarName)),
        stream_name(Stream, WinID),
pbi_write('Alias'=Alias),pbi_write(' Interp'=Interp),pbi_write(' WV'=WaitVarName),pbi_write(' WinID'=WinID),pbi_nl,
        tcl_call(Interp, [set_prompt_mark, WinID], _),
        tcl_call(Interp,
                [bind,WinID,'<Return>', [ xmit_line_plain,WinID,Alias,WaitVarName]], _),
        tcl_call(Interp, [set,WaitVarName,0], _),
        tcl_call(Interp, [wait_for_line1, WaitVarName], WaitRes),
        finish_wait_data_tcltk(WaitRes, Stream, Call).
The required binding of xmit_line_plain to <Return> occurs here.  And the pbi_write demonstrates that this wait_data call occurs when the als console is first mapped:

KB-SSD > ./alsdev
Alias=shl_tk_in_win Interp=shl_tcli WV=WaitForLine.topals.text WinID=.topals.text

I think what’s needed is to drill further into *.text events together with how things work in [wish] console.tcl.
